File: python/clock/clock.py
class Clock:
    def __init__(self, hour, minute):
        self.hour, self.minute = self.normalize(hour, minute)

    def __repr__(self):
        return f"Clock({self.hour}, {self.minute})"

    def __str__(self):
        return f"{str(self.hour).zfill(2)}:{str(self.minute).zfill(2)}"

    def __eq__(self, other):
        other_hour, other_minute = self.normalize(other.hour, other.minute)
        if self.hour == other_hour and self.minute == other_minute:
            return True
        else:
            return False

    def __add__(self, minutes):
        self.minute += minutes
        self.hour, self.minute = self.normalize(self.hour, self.minute)
        return self

    def __sub__(self, minutes):
        self.minute -= minutes
        self.hour, self.minute = self.normalize(self.hour, self.minute)
        return self

    def normalize(self, hour, minute):
        # Normalize input time
        while minute >= 60 or minute < 0:
            hour += 1 if minute >= 0 else -1
            minute -= 60 if minute >= 0 else -60

        while hour >= 24 or hour < 0:
            hour -= 24 if hour >= 24 else -24
        
        return (hour, minute)

File: c/clock/clock.c
#include "clock.h"

clock_t clock_create(int hour, int minute)
{
    // Variables
    char hour_str[3] = "00";
    char minute_str[3] = "00";
    char time[MAX_STR_LEN];
    clock_t new_clock;

    while (minute < 0 || minute >= 60)
    {
        if (minute < 0)
        {
            minute += 60;
            hour -= 1;
        }
        else if (minute >= 60)
        {
            minute -= 60;
            hour += 1;
        }
    }

    while (hour < 0 || hour >= 24)
    {
        if (hour < 0)
        {
            hour += 24;
        }
        else if (hour >= 24)
        {
            hour -= 24;
        }
    }

    // Conversion to string
    sprintf(hour_str, "%02d", hour);
    sprintf(minute_str, "%02d", minute);

    // Set up string representation
    strcpy(time, hour_str);
    strcpy(time + 3, minute_str);
    time[2] = ':';
    strcpy(new_clock.text, time);
    return new_clock;
}
clock_t clock_add(clock_t clock, int minute_add)
{
    // Variables
    char hour_str[3];
    char minute_str[3];
    int hour, minute;

    // Get hours and minutes from clock
    for (int i = 0, l = strlen(clock.text); i < l; i++)
    {
        if (i < 2)
        {
            hour_str[i] = clock.text[i];
        }
        else if (i > 2)
        {
            minute_str[i - 3] = clock.text[i];
        }
    }

    // Convert to integers
    sscanf(hour_str, "%d", &hour);
    sscanf(minute_str, "%d", &minute);

    return clock_create(hour, minute + minute_add);
}

clock_t clock_subtract(clock_t clock, int minute_subtract)
{
    // Variables
    char hour_str[3];
    char minute_str[3];
    int hour, minute;

    // Get hours and minutes from clock
    for (int i = 0, l = strlen(clock.text); i < l; i++)
    {
        if (i < 2)
        {
            hour_str[i] = clock.text[i];
        }
        else if (i > 2)
        {
            minute_str[i - 3] = clock.text[i];
        }
    }

    // Convert to integers
    sscanf(hour_str, "%d", &hour);
    sscanf(minute_str, "%d", &minute);

    return clock_create(hour, minute - minute_subtract);
}
bool clock_is_equal(clock_t a, clock_t b)
{
    return strcmp(a.text, b.text) == 0;
}
